activerecord/lib/active_record/insert_all.rb

```diff
@@ -116,9 +126,10 @@ def returning
         end
 
         def conflict_target
-          return unless conflict_columns
-          sql = +"(#{quote_columns(conflict_columns).join(',')})"
-          sql << " WHERE #{where}" if where
+          index = insert_all.unique_by_index
+
+          sql = +"(#{quote_columns(index&.columns || insert_all.primary_keys).join(',')})"
```

Comment:
I think we need to set primary keys as conflict target only if `update_duplicates?` is true.

Reply:
Good catch, @vishaltelangre. Yeah, that's true. It's OK (desirable, even) to leave the **conflict target** clause off of `ON CONFLICT DO NOTHING`. Implicitly using primary keys here would cause duplicates to be skipped (by default) _only_ if they were duplicate by primary key.

Reply:
Yes, I had a version of this that looked like the below but I cut `update_duplicates?` here as connection's `build_insert_sql` already guarded with `update_duplicates?`. However, I see now that would still fuck it up for `skip_duplicates?`, right. I think this would be clearer if it was split into `conflict_target_for_skip` and `conflict_target_for_update`. I'll experiment, thanks.

```ruby
def conflict_target
  if index = insert_all.unique_by_index
    +"(#{quote_columns(index.columns).join(',')})".tap do |sql|
      sql << " WHERE #{index.where}" if index.where
    end
  elsif update_duplicates?
    "(#{quote_columns(insert_all.primary_keys).join(',')})"
  end
end
```